### PR TITLE
Add option to replace object files with stubs during linking

### PR DESCRIPTION
**Pull request by nilsdeppe**
## Proposed changes

- After linking object files into a library or executable the object files are no longer needed (except if you are frequently rebuilding, then they can be helpful) but still take up a lot of disk space. This adds two CMake flags to have the linker replace object files with empty stubs. While not recommended to be used during normal development, on CI enabling these flags reduces disk usage of the GCC6 debug target by ~2.6GB and of the clang debug targets by ~1.5GB. I think the main way to further reduce disk usage would be to switch to shared libraries.

@wthrowe this implements the stubbing that I talked with you about on the call on Wednesday. Any feedback is greatly appreciated!

### Types of changes:

- [ ] Bugfix
- [x] New feature
- [ ] Refactor

### Component:

- [ ] Code
- [ ] Documentation
- [x] Build system
- [x] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### CODE REVIEW

File: tools/WrapExecutableLinker.sh [hunk 1]
```diff
@@ -44,3 +44,19 @@ else
     "$@" -DGIT_COMMIT_HASH=$git_commit_hash -DGIT_BRANCH=$git_branch \
          "${InfoAtLink_file}"
 fi
+
+if @WRAP_EXECUTABLE_LINKER_USE_STUB_OBJECT_FILES@; then
+    # When linking, mv the object files to a temporary file (this
+    # preserves times), then use touch to create a new file empty
```
Comment (wthrowe) [inline tools/WrapExecutableLinker.sh]: Too many "file"s, I think.  Probably you want to remove this one.

File: tools/WrapExecutableLinker.sh [hunk 2]
```diff
@@ -44,3 +44,19 @@ else
...
+    # stub object file with the same times (atime and mtime) as
+    # the original object file, except that it is 0kB
+    for file in "$@"
+    do
+        if [[ $file =~ .*\.cpp\.o ]]; then
+            temp_file="${file}.tmp_to_eliminate_size"
+            mv ${file} ${temp_file}
+            touch -r ${temp_file} ${file}
+            rm ${temp_file}
```
Comment (wthrowe) [inline tools/WrapExecutableLinker.sh]: This script is at least mostly quoting things correctly, so quote the variables in this section, too.

**Review by wthrowe — CHANGES_REQUESTED** [review]

**Comment by nilsdeppe** [comment]
Thanks for the review @wthrowe! Pushed two fixups for the linker scripts

**Comment by wthrowe** [comment]
Looks good.  Squash.

**Comment by nilsdeppe** [comment]
Squashed. Thanks again @wthrowe!

**Review by wthrowe — APPROVED** [review]

File: cmake/SetupBuildInfo.cmake [hunk 3]
```diff
@@ -33,6 +33,17 @@ else()
   file(REMOVE ${CMAKE_BINARY_DIR}/tmp/Formaline.sh)
 endif()
 
+option(
+  STUB_EXECUTABLE_OBJECT_FILES
+  "Replace executable object files with stubs to reduce disk usage."
```
Comment (nilsvu) [inline cmake/SetupBuildInfo.cmake:38]: <s>add "after linking"</s>

**Review by nilsvu — CHANGES_REQUESTED** [review]
<s>please squash directly</s> actually, nvm, let's just merge this

**Review by nilsvu — APPROVED** [review]

**Review by kidder — APPROVED** [review]

**Pull request merged**